first attempt at adding turborepo post (#1)
getcaribou/eng-blog: data/blog/hello-turborepo.md

---

> title: Turborepo At Caribou

Reviewer: Come to think of it, I think a better title might be something like "Lessons Learned While Setting Up Turborepo" or "Pitfalls when Adding Turborepo To Your Project"

Author: https://github.com/getcaribou/eng-blog/commit/e737a999f97159d0839747e384fa723c95cf714e

authors: [tcarre]
date: 2022-02-10
tags: [devops]
draft: false
summary: Challenges scaling monorepo with npm workspaces, turborepo and CircleCI
---

> We at Caribou have recently adopted a new TypeScript monorepo stack for our app frontends, starting from a single package all rebuilt by Netlify at every release.

Reviewer: > starting from a single package all rebuilt by Netlify at every release

I don't quite follow this sentence. What are you trying to convey here?

Also, perhaps link to turborepo here.

Author: I ended up just cutting off that part of the sentence.

https://github.com/getcaribou/eng-blog/commit/e737a999f97159d0839747e384fa723c95cf714e

> As our number of apps and codebases grew, we decided that we wanted to:

Author: https://github.com/getcaribou/eng-blog/commit/e737a999f97159d0839747e384fa723c95cf714e

- Save time and money on ever increasing build times

>     - Build times were increasing dramatically as we went from 2 apps in a monorepo to 4,and thus needed to build & deploy each one even though we had only made a change on one of them

Reviewer: ```suggestion
    - Build times were increasing dramatically as we went from 2 apps in our monorepo to 4. The original monorepo setup would naively deploy all apps inside the project on every push to GitHub. Once we got to 4 projects, the build times got really out of hand.
```

Author: https://github.com/getcaribou/eng-blog/commit/e737a999f97159d0839747e384fa723c95cf714e

> - Enable the granular tracking of individual application deployments for our metrics monitoring

Author: https://github.com/getcaribou/eng-blog/commit/e737a999f97159d0839747e384fa723c95cf714e

- Add CircleCI as a layer between GitHub and Netlify to manage our CI pipeline
    - Our other repos were already on CircleCI, so this allowed us to unify our CI/CD process

> - Decouple builds and deployments to save time and scale further

Author: :thinking: probably. https://github.com/getcaribou/eng-blog/commit/e737a999f97159d0839747e384fa723c95cf714e

> As we’ve faced multiple roadblocks undergoing this transition, we decided to record them for the benefit of developers at Caribou or anyone else undertaking similar endeavours.

Reviewer: ```suggestion
As we’ve faced multiple roadblocks undergoing this transition to turborepo, we decided to record them for the benefit of developers at Caribou or anyone else undertaking similar endeavours.
```

Author: https://github.com/getcaribou/eng-blog/commit/e737a999f97159d0839747e384fa723c95cf714e

## Starting point and stack choice

![Directory structure](/static/images/monorepo/1642780105.png)

We started from a flat file system containing multiple apps that were all located in the project’s root folder. The directory structure itself needed work.

### Researching & The Design Document Phase

At Caribou, net-new functionality or highly complex additions to our systems must go through a design document process.

> We wrote a design document outlining our requirements and how the new stack would meet them. Our requirements were not complicated. We wanted to rebuild and deploy only those parts of the app that have changed, and add our desired checks on CircleCI. 

Author: https://github.com/getcaribou/eng-blog/commit/e737a999f97159d0839747e384fa723c95cf714e

We had a first look at monorepo management. We knew [Lerna](https://lerna.js.org) was a popular choice, but [Turborepo](https://turborepo.org) had recently been acquired by Vercel and seemed highly promising. It purported to be very fast but simpler than Lerna, and one of our engineers had a positive experience with it.

> Turborepo works with one of Yarn, npm, or pnpm workspaces. We already used npm as a package manager, so in order to keep things familiar we went with npm workspaces.

Author: already says `It purported to be very fast but simpler than Lerna, and one of our engineers had a positive experience with it.`

```bash
# Don't do that anymore
npm run start
# Do this instead
npm run --workspace blog start
# or for short
npm run -w blog start 
```

## Separating dependencies

One detail which was not immediately obvious during the transition is that the root `package.json` should only contain dev dependencies. (It doesn’t need to be all of them, either.) We initially thought we should keep common dependencies in the root package.json. This caused React errors from having multiple instances of React running. 

Another thing to note is you should never see a `package-lock.json` inside a sub-package’s folder. This means the `npm install` command was run inside it, which is incorrect! Delete the resulting `package-lock.json` as well as the `node_modules` it newly installed. When using npm workspaces, all dependencies live in the root `node_modules`.

## Import resolution after transitioning

We use webpack for our build pipeline, and found out that `webpack` was sometimes resolving modules that `tsc` couldn’t. This is problematic, as we wanted to use `tsc` for our CI checks! After experimentation, I found that imports must adhere to the following format:

> - Absolute imports from the current package must not be prefixed with the package’s name, i.e. if you are currently inside `ha-dash` you must write `import { whatever } from 'src/components` and not `import { whatever } from ha-dash/src/components`.

Author: e737a99

> - Absolute imports from the current package must not be prefixed with the package’s name, i.e. if you are currently inside `ha-dash` you must write `import { whatever } from 'src/components` and not `import { whatever } from ha-dash/src/components`.

Author: e3043b0

- Absolute imports from other packages must be written as `{package_name}/src/some_module`
    - Unfortunately we haven’t found how to skip the `/src/`  for cross-package imports yet. [This solution](https://github.com/nodejs/node/issues/14970#issuecomment-571887546) seemed promising but it causes the typescript compiler to hang for some reason.

While transitioning and changing import paths, I’ve often used Linux shell loops like the following:

```bash
# make sure your current directory is the package you wish to perform changes in
# commit your current repo state so you can undo in case of mistake!
for file in **/**.{ts,tsx}; do
  sed -i -e "s?from 'survey-manager-src/?from '?g" $file;
done
```

while in the `survey-manager` directory, I ran this command to change all instances of `from 'survey-manager-src/` to `from '`.

## Failing tests

We use `jest` for tests, and found that in order for tests to work in our setup we needed each package to contain a `babel.config.js` file including `'@babel/preset-react'`. This may be applicable to your pipeline, too!

## CircleCI

### Saving turbo cache artifacts between builds

> Turborepo stores build artifacts at `node_modules/.cache` in order to restore files which do not need to be rebuilt. In order to save those artifacts between. W

Reviewer: > In order to save those artifacts between. W

This last bit doesn't make sense.

Author: W

Author: e737a99

      - attach_workspace:
          at: .
      - run:
          name: "Deploy to netlify"
          command: ./deploy.sh ${CIRCLE_BRANCH} ${CIRCLE_SHA1}
```

### Saving dependencies between builds

While you are at it, you can cache npm dependencies between runs. The strategy is slightly different:

```yaml
install-deps:
    executor: caribou
    steps:
      - checkout
      - restore_cache:
          keys:
            - npm-deps-{{ checksum "package-lock.json" }}
            - npm-deps-
      - run:
          name: "Install Dependencies"
          command: |
            echo "Node version: $(node -v)"
            echo "npm version: $(npm -v)"
            npm install
      - save_cache:
          key: npm-deps-{{ checksum "package-lock.json" }}
          paths:
            - node_modules
      - persist_to_workspace:
          root: .
          paths:
            - node_modules
```

We use `npm-deps-{{ checksum "package-lock.json" }}` this time, to look for cached node modules from runs of *any branch* that had the same `package-lock.json`. If none is found, we simply get the latest cached `node_modules`. Then `npm install` is run anyway, so that any missing package is added.

> ## /!\ The netlify CLI cannot use same URL prefixes as automatic branch deployments

Reviewer: Could we use an actual emoji here?

Author: e737a99

[https://github.com/netlify/cli/issues/1984#issuecomment-862554734](https://github.com/netlify/cli/issues/1984#issuecomment-862554734)

If you’ve previously used automatic netlify deployments by branch, then you might be used to having URLs formatted as `{branch-name}--{site-name}.netlify.app`.

As soon as you’ve used this feature once, you can no longer use that subdomain with the Netlify CLI. We had to move to other prefixes using the Netlify CLI `--alias` option. The documentation says to “avoid” using the same prefix as branch names, but doesn’t say why... now you know! [Here is the GitHub issue about this](https://github.com/netlify/cli/issues/1984).

## Only deploying the individual apps which turbo rebuilt

> This is something which the [documentation for the netlify CLI](https://cli.netlify.com/commands/deploy) doesn't tell you, so you won't find out until you actually run it: **the netlify CLI compares the newest build's file hashes with the previous build's hashes, and requests only those files which have changed.** In other words, you can safely use the netlify CLI to trigger deployments of *all* your packages, and netlify will only ever receives those files which have changed.

Reviewer: ```suggestion
This is something which the [documentation for the netlify CLI](https://cli.netlify.com/commands/deploy) doesn't tell you, so you won't find out until you actually run it: **the netlify CLI compares the newest build's file hashes with the previous build's hashes, and requests only those files which have changed.** In other words, you can safely use the netlify CLI to trigger deployments of *all* your packages, and netlify will only ever receive those files which have changed.
```

Author: e737a99

However, if you are using something less sophisticated than netlify, here's a bash script I wrote before I realised the netlify CLI's capabilities.

```bash

> # Save the turbo output with this command:

Reviewer: What does this this script do exactly? I don't quite follow, might want to add a bit more context about it.

Author: e737a99

deploy_app() {
  app_name=$1
  # your deployment command here
}

for app in ${APPS[@]}; do
  case "$(cat ./.build_output)" in
    *"${app}:build: cache miss, executing"*) deploy_app "$app" ;;
    *"${app}:build: cache bypass, force"*) deploy_app "$app" ;;
    # Uncomment the first *) line to force deployment
    # *) deploy_app "$app" ;;
    *) echo "turbo did not rebuild $app, not deploying." ;;
  esac
done

```

And for whoever it might help, our netlify deploy function:

```bash
# Those environment variables are set in CircleCI
site_id_of() {
  case "$1" in
    ha-dash) echo "$HA_DASH_NETLIFY_ID" ;;
    fa-dash) echo "$FA_DASH_NETLIFY_ID" ;;
    planner) echo "$PLANNER_NETLIFY_ID" ;;
    survey-manager) echo "$SURVEY_MANAGER_NETLIFY_ID" ;;
  esac
}

deploy_app() {
  app_name=$1
  if [ "$BRANCH" = "production" ]; then
    branch_option=""
  else
    branch_option="--alias staging-branch"
  fi
  # --prod argument applies to staging too
  npx netlify deploy \
    --auth=$NETLIFY_AUTH_TOKEN \
	  --dir=./apps/$app_name/build \
    --message="$BRANCH deployment of $GIT_HASH" \
    --prod \
    --site=$(site_id_of "$appName") \
    $branch_option
}
```

# Conclusion

> Do you have experience transitioning to monorepo management tools? Do you see anything we can improve? Let us know! I hope this log of some of the challenges making the transition can be helpful to some of you. Happy hacking!

Reviewer: Please add a "did you enjoy this post? we're hiring" section to the end of this blog post:

https://github.com/getcaribou/eng-blog/blob/master/data/blog/2021-review.md#did-you-enjoy-this-post-were-hiring

<img width="839" alt="image" src="https://user-images.githubusercontent.com/10621548/153673998-fa80d9b7-3930-40b9-a404-9a8d68c11d6e.png">

Author: https://github.com/getcaribou/eng-blog/commit/e737a999f97159d0839747e384fa723c95cf714e
